### PR TITLE
Begin to clean up function evaluation

### PR DESCRIPTION
**Pull request by ttuegel**
This pull request begins to clean up function evaluation, in particular:

- Hard-coding matching instead of unification.
- Removing `MonadUnify` from the matcher so that the type of the match result is constrained (e.g. no branching).

See also: #1407 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Step/EquationalStep.hs [hunk 1]
```diff
@@ -353,22 +364,93 @@ See also: 'applyRewriteRule'
...
     initial
     -- Wrap the rules so that unification prefers to substitute
     -- axiom variables.
     (map EqualityPattern.toAxiomVariables -> rules)
-  = unifyRules unificationProcedure sideCondition initial rules
+  = unifyRules sideCondition initial rules
     >>= finalizeRulesSequence sideCondition initial
+
+-- |Unifies/matches a list a rules against a configuration. See 'unifyRule'.
+unifyRules
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/EquationalStep.hs]: Here and below: why `unify` and not `match`?

File: kore/src/Kore/Step/EquationalStep.hs [hunk 2]
```diff
@@ -353,22 +364,93 @@ See also: 'applyRewriteRule'
...
+    -- ^ Top level condition.
+    -> Pattern variable
+    -- ^ Initial configuration
+    -> rule variable
+    -- ^ Rule
+    -> unifier (UnifiedRule variable (rule variable))
+unifyRule sideCondition initial rule = do
+    let (initialTerm, initialCondition) = Pattern.splitTerm initial
+        mergedSideCondition =
+            sideCondition `SideCondition.andCondition` initialCondition
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/EquationalStep.hs]: Not relevant for this PR, but we can probably remove this merge and get a performance improvement. Or make a special case for `andCondition` with top in which we don't update the representation.

**Comment by ttuegel** [comment]
I am holding off on merging this for a moment because I noticed the integration tests taking longer than normal. I am checking locally that it's due to the load on Jenkins and not a performance regression.

**Comment by ttuegel** [comment]
I found no significant change in performance locally.

**Pull request merged**